### ISSUE TITLE
OSS-Fuzz 53394: Timeout in zeek-ftp-fuzzer

### ISSUE THREAD

**Issue by timwoj** [zeek/zeek#2610]
Fuzzer: zeek-ftp-fuzzer
Detailed report: https://oss-fuzz.com/testcase-detail/5882179016327168
Minimized test case: [clusterfuzz-testcase-minimized-zeek-ftp-fuzzer-5882179016327168.zip](https://github.com/zeek/zeek/files/10077122/clusterfuzz-testcase-minimized-zeek-ftp-fuzzer-5882179016327168.zip)
Pcap file: [53394.pcap.zip](https://github.com/zeek/zeek/files/10077129/53394.pcap.zip)

```
==3289== ERROR: libFuzzer: timeout after 61 seconds
    #0 0x53b751 in __sanitizer_print_stack_trace /src/llvm-project/compiler-rt/lib/asan/asan_stack.cpp:87:3
    #1 0x45a268 in fuzzer::PrintStackTrace() /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerUtil.cpp:210:5
    #2 0x43eeb9 in fuzzer::Fuzzer::AlarmCallback() /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:301:5
    #3 0x7fc4587dd3bf in libpthread.so.0
    #4 0x4c6438 in AddressIsPoisoned /src/llvm-project/compiler-rt/lib/asan/asan_mapping.h:377:21
    #5 0x4c6438 in __asan::QuickCheckForUnpoisonedRegion(unsigned long, unsigned long) /src/llvm-project/compiler-rt/lib/asan/asan_interceptors_memintrinsics.h:31:13
    #6 0x4d1e64 in __interceptor_vsnprintf /src/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_common_interceptors.inc:1665:1
    #7 0x4d3430 in __interceptor_snprintf /src/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_common_interceptors.inc:1736:1
    #8 0x7fc45c89b4e4 in zeek::detail::HashKey::Describe(zeek::ODesc*) const zeek/src/Hash.cc:266:4
    #9 0x7fc45c89af0c in zeek::detail::HashKey::Hash() const zeek/src/Hash.cc:217:2
    #10 0x7fc45ca99f99 in zeek::Dictionary<zeek::TableEntryVal>::Lookup(zeek::detail::HashKey const*) const zeek/src/include/zeek/Dict.h:562:47
    #11 0x7fc45ca99f99 in zeek::TableVal::Find(zeek::IntrusivePtr<zeek::Val> const&) zeek/src/Val.cc:1874:34
    #12 0x7fc45c7290dc in zeek::detail::InExpr::Fold(zeek::Val*, zeek::Val*) const zeek/src/Expr.cc:4564:33
    #13 0x7fc45c6bc7df in zeek::detail::BinaryExpr::Eval(zeek::detail::Frame*) const zeek/src/Expr.cc:769:9
    #14 0x7fc45c9c96cd in zeek::detail::ExprStmt::Exec(zeek::detail::Frame*, zeek::detail::StmtFlowType&) zeek/src/Stmt.cc:433:14
    #15 0x7fc45c9e0f41 in zeek::detail::StmtList::Exec(zeek::detail::Frame*, zeek::detail::StmtFlowType&) zeek/src/Stmt.cc:1697:23
    #16 0x7fc45c85c670 in zeek::detail::ScriptFunc::Invoke(std::__1::vector<zeek::IntrusivePtr<zeek::Val>, std::__1::allocator<zeek::IntrusivePtr<zeek::Val> > >*, zeek::detail::Frame*) const zeek/src/Func.cc:416:25
    #17 0x7fc45c617a60 in zeek::EventHandler::Call(std::__1::vector<zeek::IntrusivePtr<zeek::Val>, std::__1::allocator<zeek::IntrusivePtr<zeek::Val> > >*, bool) zeek/src/EventHandler.cc:105:10
    #18 0x7fc45c6137bf in zeek::Event::Dispatch(bool) zeek/src/Event.cc:56:12
    #19 0x7fc45c614edd in zeek::EventMgr::Drain() zeek/src/Event.cc:157:13
    #20 0x572793 in LLVMFuzzerTestOneInput zeek/src/fuzzers/generic-analyzer-fuzzer.cc:75:19
    #21 0x4404e3 in fuzzer::Fuzzer::ExecuteCallback(unsigned char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerLoop.cpp:611:15
    #22 0x42bc42 in fuzzer::RunOneTest(fuzzer::Fuzzer*, char const*, unsigned long) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:324:6
    #23 0x4314ec in fuzzer::FuzzerDriver(int*, char***, int (*)(unsigned char const*, unsigned long)) /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerDriver.cpp:860:9
    #24 0x45aa22 in main /src/llvm-project/compiler-rt/lib/fuzzer/FuzzerMain.cpp:20:10
    #25 0x7fc4585d70b2 in __libc_start_main /build/glibc-eX1tMB/glibc-2.31/csu/libc-start.c:308:16
    #26 0x421e0d in _start
```

**Comment by timwoj** [comment]
Relevant comment from @awelzel in zeek-security issue:

The pcap takes ~40sec to process with an ASAN built.

A FTP command with a cmd size of 558298 bytes is queued into pending commands during `ftp_request`. That is followed by a few hundred `ftp_reply` messages, which call `get_pending_command()` over and over again without actually removing the command - it would be removed during a new `ftp_request`.

The following two places are hashing this 558kb of string over and over again for the `InExpr`:

https://github.com/zeek/zeek/blob/b9a4afe9bfedf03615a830d1d15ec8547a1543fd/scripts/base/protocols/ftp/utils-commands.zeek#L98-L101

https://github.com/zeek/zeek/blob/b9a4afe9bfedf03615a830d1d15ec8547a1543fd/scripts/base/protocols/ftp/main.zeek#L316-L319

A 550KB long FTP command is obviously weird (32 chars would already be weird) and maybe an `AnalyzerViolation()` or truncation is the right answer here (maybe at length 1024), but it clearly falls into the same discussions we had in #113 for SMTP.


Here's a flamegraph when running zeek -r.
![zeek-r](https://user-images.githubusercontent.com/2100509/201744821-7e22251b-79a9-497b-bef0-9f82f5e2eca5.svg)

**Comment by timwoj** [comment]
Fixed in master as 3f5cb75a2a11279acd1b60bbdb35f95a47ad1089
